src/format.rs

```diff
@@ -0,0 +1,115 @@
+use crate::PrintFmt;
+use core::ffi::c_void;
+use core::fmt::{self, Write};
+
+/// Inspects the current call-stack, formatting it to `stream`.
+///
+/// This is similar to `write!(stream, "{:?}", Backtrace::new())`,
+/// but avoids memory allocation.
+/// This can be useful in a signal handler, where allocation may cause deadlocks.
```

Comment:
This unfortunately technically isn't quite true because the underlying implementation almost always allocates memory in one way or another. While this avoids all Rust-based allocations, the backtrace unwinder/etc, if called for the first time will allocate memory here.

In general nothing in this crate is suitable for usage in a signal handler right now unfortunately.

Reply:
I see.

For what it’s worth using something like this PR instead of `Backtrace::new()` in Servo’s signal handler does seem to fix a deadlock, presumably through reducing the number of allocations. (Or at least I couldn’t reproduce it after running the test case a hundred times, whereas with `Backtrace::new()` I would reliably reproduce within ten runs.) But it’s quite possible that there’s still another deadlock waiting to happen and maybe we should remove that handler entirely.

I’ll remove mention of signal handler in these doc-comments. What else do you think could be a good use case for this new feature? In theory this is doing less work and might be more efficient than creating a `Backtrace` only to print it and immediately drop it, but I don’t know if the difference has a practical impact.

Reply:
Yes it's just a deadlock waiting to happen. The underlying implementation of backtraces are not at all guaranteed to not malloc, but they try not to to be efficient. `Backtrace::new()` is guaranteed to allocate, where this API as-is just allocates some of the time, not all the time. 

Getting a backtrace from a signal handler is a really tricky thing to do and generally not covered by this crate.

TBH I'm not sure what a use case for this feature is, folks are generally ok with the mild cost that `Backtrace::new` has

Reply:
Right, my motivation for this PR was not the CPU time cost but trading a deadlock I could reliably reproduce for a possible one I haven’t seen happen. (Knowing this still isn’t a fully correct solution.)

If the only use case for this is incorrect and something the project doesn’t want to cover, maybe it shouldn’t be included? Even though based on what this code does it seems like within this crate would be the right location.

I’m OK with closing this PR in that case.